Python/clinic/instrumentation.c.h

```diff
@@ -0,0 +1,311 @@
+/*[clinic input]
+preserve
+[clinic start generated code]*/
+
+#if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
+#  include "pycore_gc.h"            // PyGC_Head
+#  include "pycore_runtime.h"       // _Py_ID()
+#endif
+
+
+PyDoc_STRVAR(monitoring_use_tool_id__doc__,
+"use_tool_id($module, tool_id, name, /)\n"
+"--\n"
+"\n");
+
+#define MONITORING_USE_TOOL_ID_METHODDEF    \
+    {"use_tool_id", _PyCFunction_CAST(monitoring_use_tool_id), METH_FASTCALL, monitoring_use_tool_id__doc__},
```

Comment:
The spec is that registering IDs is a politeness, not a requirement.
The value of having IDs is that well behaved tools don't accidentally trample on each other.

Forcing tools to register might give an illusion of ownership.
There is no way to check that the tool that registered the ID is the same one as is setting events, nor is there anyway to stop a tool taking another's ID.

Reply:
But checking that the ID is registered helps everyone stay polite. These IDs are like file descriptors. I like to require registration.

Reply:
(Come to think of it, maybe it could work like open() and return the next available ID.)

Reply:
> (Come to think of it, maybe it could work like open() and return the next available ID.)

There are recommended IDs for tools to use. The idea being that you might want a debugger and coverage tool, but you almost certainly don't want two debuggers.

Reply:
I see your point, but this is a change to the PEP, with all the rigmarole that involves.

The code change should be simple enough.

Reply:
The predefined IDs could be like pre-opens FDs, e.g. 0, 1, 2 for std(in,out,err). I’ll take it to the SC as well.

Reply:
+1 on the approach of asking for the next available id.